src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Infer.java

```diff
@@ -651,6 +654,90 @@ public Type instantiateFunctionalInterface(DiagnosticPosition pos, Type funcInte
             return owntype;
         }
     }
+
+    public Type instantiatePatternType(Type expressionType, TypeSymbol patternTypeSymbol) {
+        if (expressionType.tsym == patternTypeSymbol)
```

Comment:
not sure we are dealing with cases when expressionType is a type variable or an intersection type, are there tests covering those cases, sorry if I'm missing something here

Reply:
Good catch, thanks! Should be fixed in [5afc602](https://github.com/openjdk/jdk/pull/10814/commits/5afc60236005d5ea77984f2baeca7e928334ce3a). Thanks!